Reqnroll/Formatters/PubSub/CucumberMessagePublisher.cs

```diff
@@ -0,0 +1,441 @@
+
+    // The following methods handle the events published by the TestThreadExecutionEventPublisher
+    // When one of these calls the Broker, that method is async; otherwise these are sync methods that return a completed Task (to allow them to be called async from the TestThreadExecutionEventPublisher)
+    private async Task FeatureStartedEventHandler(FeatureStartedEvent featureStartedEvent)
+    {
+        var featureInfo = featureStartedEvent.FeatureContext?.FeatureInfo;
+
+        //if (!_startupCompleted)
+        //    throw new InvalidOperationException($"Formatters attempting to start processing on feature {featureInfo.Title} before the message publisher is ready.");
+
+        if (!_enabled || featureInfo == null)
+        {
+            return;
+        }
+
+        // The following should be thread safe when multiple instances of the Test Class are running in parallel. 
+        // If _startedFeatures.ContainsKey returns true, then we know another instance of this Feature class has already started. We don't need a second instance of the 
+        // FeatureExecutionTracker, and we don't want multiple copies of the static messages to be published.
+        if (_startedFeatures.ContainsKey(featureInfo))
+        {
+            // Already started, don't repeat the following steps
+            return;
+        }
+
+        var featureExecutionTracker = new FeatureExecutionTracker(featureStartedEvent, _testRunStartedId, SharedIdGenerator, StepDefinitionIdByMethodBinding, _messageFactory);
+        if (_startedFeatures.TryAdd(featureInfo, featureExecutionTracker) && featureExecutionTracker.Enabled)
+        {
+            try
+            {
+                foreach (var msg in featureExecutionTracker.StaticMessages) // Static Messages == known at compile-time (Source, Gherkin Document, and Pickle messages)
+                {
+                    await _broker.PublishAsync(msg);
+                }
+            }
+            catch (System.Exception ex)
+            {
+                _logger.WriteMessage($"Error publishing messages: {ex.Message}");
+            }
+        }
+    }
+
+    internal Task FeatureFinishedEventHandler(FeatureFinishedEvent featureFinishedEvent)
+    {
+        // For this and subsequent events, we pull up the FeatureExecutionTracker by featureInfo.
+        var featureInfo = featureFinishedEvent.FeatureContext?.FeatureInfo;
+        if (!_enabled || featureInfo == null)
+        {
+            return Task.CompletedTask;
+        }
+        if (!_startedFeatures.ContainsKey(featureInfo) || !_startedFeatures[featureInfo].Enabled)
+        {
+            return Task.CompletedTask;
+        }
+        var featureTracker = _startedFeatures[featureInfo];
+        featureTracker.ProcessEvent(featureFinishedEvent);
+        foreach (var msg in featureTracker.RuntimeGeneratedMessages)
+        {
+            _messages.Add(msg);
+        }
+        _runStatus = _runStatus & featureTracker.FeatureExecutionSuccess;
```

Comment:
There is a race condition here, if one thread completes the successful feature at the same time as another thread completes an unsuccessful, the successful might override the unsuccessful.
Use an if instead and only set it to false: `if (!featureTracker.FeatureExecutionSuccess) _runStatus = false;`